File: frontend/src/api/assistance.ts
import { request } from './client'
import type { ApiSchemas } from './generated'

export type AssistanceTarget = ApiSchemas['AssistanceTarget']
export type AssistanceContext = ApiSchemas['AssistanceContext']
export type AssistanceGuide = ApiSchemas['AssistanceGuide']
export type AssistanceModel = ApiSchemas['AssistanceModel']
export type AssistanceTurn = ApiSchemas['AssistanceTurn']
export type AssistanceTool = ApiSchemas['AssistanceTool']
export type AssistanceToolResult = ApiSchemas['AssistanceToolResult']
export type CommitConnectionReview = ApiSchemas['CommitConnectionReview']

export const assistanceApi = {
  reviewCommitConnection: (runId: string) =>
    request<CommitConnectionReview>(
      `/assistance/runs/${runId}/commit-connection`,
    ),
  executeTool: (payload: ApiSchemas['AssistanceToolRequest'], csrf: string) =>
    request<AssistanceToolResult>(
      '/assistance/tools/execute',
      {
        method: 'POST',
        body: JSON.stringify(payload),
      },
      csrf,
    ),
  catalog: () => request<AssistanceGuide[]>('/assistance/catalog'),
  models: () => request<AssistanceModel[]>('/assistance/models'),
  context: (target: AssistanceTarget) => {
    const query = new URLSearchParams({ zone: target.zone })
    if (target.project_id) query.set('project_id', target.project_id)
    if (target.chat_id) query.set('chat_id', target.chat_id)
    return request<AssistanceContext>(`/assistance/context?${query}`)
  },
  suggestions: (
    payload: ApiSchemas['SuggestionsRequest'],
    csrf: string,
    signal: AbortSignal,
  ) =>
    request<ApiSchemas['AssistanceSuggestions']>(
      '/assistance/suggestions',
      {
        method: 'POST',
        body: JSON.stringify(payload),
        signal,
      },
      csrf,
    ),
  send: (
    payload: ApiSchemas['AssistanceMessage'],
    csrf: string,
    signal: AbortSignal,
  ) =>
    request<ApiSchemas['AssistanceAnswer']>(
      '/assistance/messages',
      {
        method: 'POST',
        body: JSON.stringify(payload),
        signal,
      },
      csrf,
    ),
}

File: backend/src/agents_ide/api/assistance.py
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from agents_ide.api.deps import get_secret_store, get_session, get_settings
from agents_ide.config import Settings
from agents_ide.security.secrets import SecretStore
from agents_ide.services import assistance
from agents_ide.services.assistance_catalog import AssistanceGuide, AssistanceTarget, catalog
from agents_ide.services.assistance_suggestions import (
    AssistanceSuggestions,
    SuggestionsRequest,
    generate,
)
from agents_ide.services.assistance_tools import (
    AssistanceToolRequest,
    AssistanceToolResult,
    execute,
)
from agents_ide.services.commit_connection import CommitConnectionReview, review_connection

router = APIRouter(prefix="/api/assistance", tags=["assistance"])
SessionDep = Annotated[Session, Depends(get_session, scope="function")]
SettingsDep = Annotated[Settings, Depends(get_settings)]


@router.get("/catalog", response_model=list[AssistanceGuide])
def assistance_catalog() -> list[AssistanceGuide]:
    return catalog()


@router.get("/models", response_model=list[assistance.AssistanceModel])
def assistance_models(session: SessionDep) -> list[assistance.AssistanceModel]:
    return assistance.models(session)


@router.get("/context", response_model=assistance.AssistanceContext)
def assistance_context(
    session: SessionDep, settings: SettingsDep, target: Annotated[AssistanceTarget, Query()]
) -> assistance.AssistanceContext:
    return assistance.collect_context(session, settings, target)


@router.post("/messages", response_model=assistance.AssistanceAnswer)
def assistance_message(
    session: SessionDep,
    settings: SettingsDep,
    secrets: Annotated[SecretStore, Depends(get_secret_store)],
    payload: assistance.AssistanceMessage,
) -> assistance.AssistanceAnswer:
    return assistance.answer(session, settings, secrets, payload)


@router.post("/suggestions", response_model=AssistanceSuggestions)
def assistance_suggestions(
    session: SessionDep,
    settings: SettingsDep,
    secrets: Annotated[SecretStore, Depends(get_secret_store)],
    payload: SuggestionsRequest,
) -> AssistanceSuggestions:
    return generate(session, settings, secrets, payload)


@router.post("/tools/execute", response_model=AssistanceToolResult)
def assistance_tool(session: SessionDep, payload: AssistanceToolRequest) -> AssistanceToolResult:
    return execute(session, payload)


@router.get("/runs/{run_id}/commit-connection", response_model=CommitConnectionReview)
def assistance_commit_connection(session: SessionDep, run_id: str) -> CommitConnectionReview:
    return review_connection(session, run_id)
